Migrate to typescript

Darkflame72 opened the pull request:
Migrate to typescript to take advantage of type hints and documentation generation.

extremeheat commented on package.json:
```diff
@@ -2,16 +2,39 @@
   "description": "Provide easy access to minecraft data in node.js",
-  "main": "index.js",
-  "tonicExampleFilename": "example.js",
+  "main": "build/main/index.js",
```
Don’t think anything needs to change here other than running tsc and switching to standard for typescript

extremeheat commented on src/indexes.ts:
```diff
@@ -0,0 +1,101 @@
+import { indexer } from './indexer';
+
+module.exports = function (mcData: {
+  readonly biomes: readonly any[] | undefined;
```
Too much duplicative code here. It might make sense to define this as an interface in a separate file so it doesn’t pollute the logic. Also, the type data here can probably be generated from the json schemas

extremeheat commented on src/indexer.ts:
```diff
@@ -0,0 +1,76 @@
+export const indexer = {
+  buildIndexFromObject: function (
+    object: { readonly [x: string]: any } | undefined,
```
are all the type annotations here really necessary? it seems excessive here, personally it reduces readability of the code

extremeheat reviewed (commented):
Looking through the code, I'm not quite if there's much benefits to doing typescript here, the library is relatively small as-is and pretty much all of the important type related definitions already exist (although in separate files).

Karang commented:
As I'm personally not doing typescript, I liked it better when it was limited to js + ts typing. What is the benefit of doing this changes?

Typescript noob question: I thought typescript was about adding type safety to js, isn't giving almost every variable the type `any` or `unknown` defeating that purpose ?

rom1504 commented on package.json:
```diff
@@ -2,16 +2,39 @@
-    "fix": "standard --fix"
+    "build": "run-p build:*",
+    "build:main": "tsc -p tsconfig.json",
+    "build:module": "tsc -p tsconfig.module.json",
```
We definitely want to keep using standard not prettier
Look at the doc for how to use it

Darkflame72 replied:
It looks like standard does not support typescript so we could use eslint?

extremeheat replied:
See how it’s done here - https://github.com/PrismarineJS/mineflayer-statemachine/blob/standard-impl/package.json

Darkflame72 replied:
Yep I can move it to the ts-standard plugin

rom1504 commented on package.json:
```diff
@@ -2,16 +2,39 @@
+    "test:spelling": "cspell \"{README.md,.github/*.md,src/**/*.ts}\"",
+    "test:unit": "nyc --silent ava",
+    "check-cli": "run-s test diff-integration-tests check-integration-tests",
+    "check-integration-tests": "run-s check-integration-test:*",
```
This is way too much boilerplate

Darkflame72 replied:
Yep just had some base typescript conversions which I made and am just removing unnecessary features now

rom1504 commented on package.json:
```diff
@@ -33,12 +56,68 @@
+    "@ava/typescript": "^1.1.1",
+    "@istanbuljs/nyc-config-typescript": "^1.0.1",
+    "@types/node": "^14.14.7",
+    "@typescript-eslint/eslint-plugin": "^4.0.1",
```
This is way too much boilerplate

rom1504 commented:
For context, we used to use babel in all repos to use new js (es2015 when it was not available). I'm pretty happy it's now gone, it's was a huge amount of boilerplate.
Can we try to avoid boilerplate as most as possible here ?

The criteria for introducing typescript are :
1. Not adding too much boilerplate
2. Significant benefits in term of code quality

Darkflame72 commented:
Yep this is not finished and some of it was me trying different things to fully understand it hence the draft status

Darkflame72 commented:
> As I'm personally not doing typescript, I liked it better when it was limited to js + ts typing. What is the benefit of doing this changes?
> 
> 
> 
> Typescript noob question: I thought typescript was about adding type safety to js, isn't giving almost every variable the type `any` or `unknown` defeating that purpose ?

Yeah that does defeat the purpose but I haven't finished adding the types and when I convert they are auto generated to any so I have to manually update to their actual type which I am doing

Darkflame72 commented on src/loader.ts:
```diff
@@ -0,0 +1,150 @@
+export function mcDataToNode(mcData: {
+  readonly blocks: any;
```
Yep I will do

Darkflame72 commented:
I have removed a lot of boilerplate code and added all the type hints so just need to make them all the right type and test it and do a little cleanup

rom1504 commented:
Bugs:
* You deleted the submodule
* You deleted the generation of types from schemas

Comments on current state:
The only gain currently is to use es6 export syntax right ? Anything else ?

Darkflame72 commented:
> Bugs:
> 
> * You deleted the submodule
What submodule?

> * You deleted the generation of types from schemas
The generation of types from schemas was just generating it with type of any so was not doing anything. This will generate types off the code which is more accurate especially when their are changes
> 
> Comments on current state:
> The only gain currently is to use es6 export syntax right ? Anything else ?
es6 export syntax is an advantage.
So is static typing which even as I am not finished is generating better type declarations
readability
predictability

There is no disadvantage to moving to typescript and only advantages

rom1504 commented:
Not talking about theoretical advantage. I mean directly in this pull request.
About disadvantages yes there are a bunch but they are not specific to this pull request (more tooling, harder for beginners to contribute, less stable language)

rom1504 commented:
The minecraft-data submodule where data is coming from, look at master

rom1504 commented:
Where are you reading the schemas to generate better types ?

Darkflame72 commented:
> The minecraft-data submodule where data is coming from, look at master

that was not intentional fixing now

Darkflame72 commented:
> Where are you reading the schemas to generate better types ?

https://github.com/PrismarineJS/node-minecraft-data/blob/master/typings/index-template.d.ts

Darkflame72 commented:
> Not talking about theoretical advantage. I mean directly in this pull request.
> About disadvantages yes there are a bunch but they are not specific to this pull request (more tooling, harder for beginners to contribute, less stable language)

This pull request is not finished but the advantages will be better type definitions, easier for someone to work with the current code to make improvements/modifications, es6 syntax

Darkflame72 commented:
If you don't want typescript and want to stick with what you have that is up to you but if so can you please tell me.

rom1504 commented:
As I said from the beginning I want type script if it's possible with minimal boilerplate and significant benefits.
Let's see if that happens here.

Darkflame72 commented:
whoops, accidentally pressed the wrong button, this has working typescript with all the types and no boilerplate

rom1504 commented:
There are at least those to fix:
* Typo in interface file name
* Ci is not passing
* Type generation for the data (from schema) was removed

rom1504 commented:
Ah yeah and I just saw the main problem actually: the example file should not change at all. No breaking change is acceptable for this PR.
Is that possible ?

rom1504 commented:
But most importantly, can you type out the benefits that have been added in this PR ? I looked through the code carefully and I'm not seeing it. What error will be avoided thanks to this change ?

Darkflame72 commented:
Mostly the advantage of having good type hints from anyone who uses this library in the future as it currently doesn't have any good type hints, it should also make any future changes easier and simpler to debug and for people not familiar with the code to use it.

In terms of the example file and the breaking change es6 has changed how default exports work while you used to be able to do `module.exports = function(a, b)` and then `require("name")(a, b)` es6 syntax doesn't really support default and actively discourages them for a number of reasons so naming the function would be the simple and recommended way but it would be a breaking change.

I will fix the typo and ci will take a little while as ts-standard is super super strict, the type generation from schemas is no longer needed as the types are now done in the interfaces.ts file so they are directly referenced in the code and allow for much better types

rom1504 commented:
Ok I guess that makes sense overall.
Is there really no way to avoid the breaking change ?
I don't think we can afford it. This module is used by about 15 prismarinejs modules and hundred of users code. We cannot change it all to adapt to breaking changes.
And this .default doesn't seem like a good change from an user point of view no ?

rom1504 commented:
If it's really impossible to not have a breaking change due to es6 export, I think it should be reverted to commonjs module.exports

Darkflame72 commented:
the .default is not great at all and best would be to move to a named function or we could go back to module.exports and that would be the way to have no breaking changes and has no disadvantages that I can think of.

Darkflame72 commented:
The only advantage is being able to only import what you want instead of importing everything which module.exports does

rom1504 commented:
Is it possible to have both somehow ? module.exports to keep the current thing working and an es6 export in another file or something such that it's possible to use it if needed

Darkflame72 commented:
I will take a look but I don't think so, it is up to you which system you want, the main advantage of import and export is less memory use so not massive of an advantage

rom1504 commented:
Ok then let's keep compatibility and use commonjs

Darkflame72 commented:
fixed file name and commonjs exports so will fix ci over the next day, everything currently does work

rom1504 commented:
Is there not a ts-standard --fix ?

TheDudeFromCI commented:
Yes, ts-standard has a --fix.

DeltaEvo commented:
> Mostly the advantage of having good type hints from anyone who uses this library in the future as it currently doesn't have any good type hints, it should also make any future changes easier and simpler to debug and for people not familiar with the code to use it.

I don't understand the issue with the current typing of this library ? The only issue that I see is that the `IndexedData` type have dynamic keys where It could be static if we enhance the generation

I understand that typing is easier when your codebase is in typescript but since the library here is exposing generated data, generating type seems to be the best thing to do here for me

From what I see with this PR the typing quality seems to have decreased instead, `build/index.d.ts` is empty :thinking: I think this is a bug but I don't know a lot about typescript

All things with a string key are typed now with the `any` type (`blocks`, `blocksByName`, ...) (return type of `mcDataToNode` found in `build/loader.d.ts`)

There is a new `interfaces.ts` that must be handwritten and is currently incomplete

Darkflame72 commented:
> Is there not a ts-standard --fix ?

The issue is with typings not with the actual code so has to be fixed manually

Darkflame72 commented:
Well this is fully working and CI is failing not due to code quality but how ci works.

DeltaEvo commented on src/interfaces.ts:
```diff
@@ -0,0 +1,1156 @@
+/* eslint-disable */
```
This seems to have been generated, maybe with quicktype ?
Can we not generate it as part of the build ?

Darkflame72 replied:
if we disable it we have around 100 errors thrown by standard-ts which will cause everything to fail even though it is valid ts

DeltaEvo replied:
My comment is not on the eslint-disable but more on the whole file

Darkflame72 replied:
we could generate the file pre-build and not commit it to git like we do to data?

DeltaEvo replied:
Yes I think it whould be better, because this also mean that we don't need to update it if the schema changes

Darkflame72 replied:
I will finish that in the morning as it's causing some issues right now and I need sleep

DeltaEvo commented:
The ci is using Github CI check `.github/workflows/ci.yml` file

Darkflame72 commented:
> The ci is using Github CI check `.github/workflows/ci.yml` file

Yep am fixing it now to deal with typescript

Darkflame72 commented:
This should be all finished now

Darkflame72 commented:
@rom1504 are there any issues with this?

rom1504 commented:
Yes the code is strictly worse after this PR.

Darkflame72 commented:
> Yes the code is strictly worse after this PR.

In what way? I have added typings which you did not have before and it improves stability of your code.

Darkflame72 commented:
Should I close this or is there still interest?
